### ISSUE TITLE
Closing command line Wine application causes all other Wine sessions to freeze

### ISSUE THREAD

**Issue by sugar700** [flathub/org.winehq.Wine#41]
If you have a graphical Wine applications open, closing a text application will cause Wine applications to freeze.

For instance, you can run `winecfg`, followed by `cmd`, and then exit `cmd` by typing `exit`, and then `winecfg` will freeze.

**Comment by fastrizwaan** [comment]
`flatpak kill org.winehq.Wine` would help to close all instances.

**Comment by xalt7x** [comment]
Sounds similar to https://github.com/bottlesdevs/Bottles/issues/2759

**Comment by foresto** [comment]
Reproducible with two instances of winecfg:

1. From terminal A, `flatpak run --command=winecfg org.winehq.Wine`
2. From terminal B, `flatpak run --command=winecfg org.winehq.Wine`
3. Close the second instance of winecfg

Both terminal A and B have now returned to their shell prompts. The first winecfg window is still present, but is now frozen.

**Comment by charlesthobe** [comment]
~~Might also be related to this: https://github.com/bottlesdevs/Bottles/issues/2372~~
albeit `sudo flatpak override --allow=per-app-dev-shm org.winehq.Wine` doesn't fix it.

**Comment by charlesthobe** [comment]
Thing is if you run `flatpak run --command=sh org.winehq.Wine` then run `winecfg &` and then `wine notepad &` etc, the bug doesn't happen, so I think the bug has something to do with creating multiple bwrap instances.

**Comment by charlesthobe** [comment]
The process space of one instance doesn't get exposed to another, this is most likely what causes the issue because of problems with IPC.
IDK if it's possible to create an intermediary executable that detects if an instance of the application is already running to launch the new command inside that instance, but if it's possible then it's probably the solution.
https://github.com/flatpak/flatpak/issues/1963

**Comment by KOPRajs** [comment]
Is there still no way to do this?

For me, if I open 2 apps and then close any of them, it closes both. I think it is caused by the same issue.

If we could launch all apps in a single instance of the flatpak container, it would probably solve most issues.

**Comment by KOPRajs** [comment]
I believe that with the help of the developers from the Wine's IRC channel we've fully analyzed this issue.

Here is the breakdown for a reference:

When the first Wine container is launched, the application starts an instance of the _wineserver_ in its container / namespace and the _wineserver_ creates a shared socket for IPC between all Wine processes in `/tmp/.wine-1000`.

This is an example process list in the first Wine container with App1:
```
~$ flatpak enter 1 /usr/bin/ps ax
    PID TTY      STAT   TIME COMMAND
      1 ?        S      0:00 bwrap --args 43 -- wine C:\Program Files (x86)\App1\app1.exe
      2 ?        S      0:00 wine C:\Program Files (x86)\App1\app1.exe
      4 ?        Ss     5:15 /app/bin/wineserver
     10 ?        Ssl    0:00 C:\windows\system32\services.exe
     13 ?        Ssl    0:29 C:\windows\system32\winedevice.exe
     22 ?        Ssl    0:00 C:\windows\system32\plugplay.exe
     24 ?        Ssl    0:00 C:\windows\system32\explorer.exe /desktop
     30 ?        Ssl    0:00 C:\windows\system32\svchost.exe -k LocalServiceNetworkRestricted
     41 ?        Ssl   11:43 C:\windows\system32\winedevice.exe
     63 ?        Ssl    0:00 C:\windows\system32\rpcss.exe
     74 pts/3    R+     0:00 /usr/bin/ps ax
```

Note, that the app1.exe has the PID 2.

When the second application is started, it connects to the shared socket in `/tmp/.wine-1000` (yes, this socket is shared between the containers) and so it does not need to start its own instance of the _wineserver_ and instead connects to the already running instance in the first container.

This is an example process list in the second container with App2:
```
~$ flatpak enter 2 /usr/bin/ps ax
    PID TTY      STAT   TIME COMMAND
      1 ?        S      0:00 bwrap --args 43 -- wine C:\Program Files (x86)\App2\app2.exe
      2 ?        S      0:00 wine C:\Program Files (x86)\App2\app2.exe
      4 pts/3    R+     0:00 /usr/bin/ps ax
```

Note, that the app2.exe has also the PID 2 in its namespace.

Connecting to the _wineserver_ via the shared socket is in general a good thing, because running 2 instances of _wineserver_ in the same Wine prefix is not supported and it is expected to cause issues with overwriting registry files. This way we get 2 apps running both connected to a single instance of _wineserver_ in the container with the App1.

Up until now everything is fine, 2 apps are running connected to a single _wineserver_.

The issue:
When the App2 is closed, it sends a message using the shared socket to the _wineserver_ saying that the Windows app with a PID 2 is about to close and terminates itself. The _wineserver_ makes a check whether the app with PID 2 is really terminated and if not, it sends a kill signal to the PID 2 and kills the App1 in its own namespace, which happens to have the same PID as App2.

The solution:
There were some proposed solutions, but none of them is easy to do.

1. Share the PID namespace between the containers.
Unfortunately, it seems impossible to do in Flatpak.

2. Run everything in one instance of the container.
May be possible to do using `flatpak enter` instead of `flatpak run`, but not easy. It seems that `flatpak enter` can't properly set the ENV variables the same way as `flatpak run` does.

3. Somehow offset the PIDs in the first container, so they do not overlap with the other containers.

4. Create a patch to disable the kill signal in the _wineserver_ for the Flatpak version.
This is a little hacky, but it might be the best option.

Unfortunately, there is now a new MR for the Wine, which would prevent the issue by refusing to run the second application at all, making it even worse: https://gitlab.winehq.org/wine/wine/-/merge_requests/9553
If it is accepted in upstream, the Wine 11.0 in Flatpak might need to revert this.

**Comment by KOPRajs** [comment]
Also, there is a workaround for the current version.

1. Run App1.

2. Run App2.

3. Close App1.
By closing the App1 in the correct namespace of the _wineserver_ first, we get rid of the PID 2 in this namespace. The container itself remains active until the _wineserver_ is closed (until all Windows apps are closed).

You can now run and close as many apps as you want, because they all share the PID 2 and there is no process with PID 2 in the namespace of the _wineserver_. It works until you close all Windows apps and the _wineserver_ is terminated (and the container of the first app is closed with it).

**Comment by Erick555** [comment]
I think preventing running multiple apps in single prefix at same time may be better outcome for users than freezing them some time later which may result in lost work. It doesn't fix the issue but it make it visible from the beginning before it does harm.

**Comment by KOPRajs** [comment]
I agree, but it would greatly limit the usability of the Wine in Flatpak. We only have single prefix, which means to only run one app at a time (even only one instance of the same app).

I don't know anything about the Flatpak build process, but can we apply a small patch to the source code? If so and no better solution is found, then I would recommend to apply 2 small patches:
1. revert the linked MR, if it is merged for the Wine 11.0
2. comment out the kill signal send from the _wineserver_

That should solve the issue and allow as to run multiple apps.

**Comment by foresto** [comment]
> I agree, but it would greatly limit the usability of the Wine in Flatpak. We only have single prefix, which means to only run one app at a time 

Wouldn't that only be true if you're treating the Wine flatpak as a stand-alone application? Is that what you're doing? It seems to me that would create more problems than just the one discussed here, such as breaking isolation between the applications that run in Wine.

Why not treat Wine as a runtime, or a base application upon which flatpak apps can be built? This way, each app would have its own wine prefix and wineserver instance. I do this with `base: org.winehq.Wine` in the manifest of an app that I package myself.

**Comment by Alcaro** [comment]
To make things worse, process killing isn't the only symptom of this issue. For example, ReadProcessMemory() and GetProcessMemoryInfo() will return nonsense, or answers corresponding to wrong process. Demo program:

```c
#include <windows.h>
#include <stdio.h>
#include <psapi.h>

int main()
{
	HANDLE h; // pretend it's a different process (the exact value GetCurrentProcess() is special cased in ReadProcessMemory())
	DuplicateHandle(GetCurrentProcess(), GetCurrentProcess(), GetCurrentProcess(), &h, 0, TRUE, DUPLICATE_SAME_ACCESS);
	
	char foo[7] = "walrus";
	char bar[7] = "nope:(";
	size_t tmp = -1;
	BOOL ok = ReadProcessMemory(h, foo, bar, 7, &tmp);
	printf("%d %d %zd %s\n", ok, GetLastError(), tmp, bar);
	
	PROCESS_MEMORY_COUNTERS pmc;
	pmc.PeakWorkingSetSize = 123;
	ok = GetProcessMemoryInfo(h, &pmc, sizeof(pmc));
	printf("%d %d %zu\n", ok, GetLastError(), pmc.PeakWorkingSetSize);
}
```
```
$ x86_64-w64-mingw32-gcc a.c && wine a.exe
1 0 7 walrus
1 0 7667712
$ x86_64-w64-mingw32-gcc a.c && unshare -Upf wine a.exe
0 5 0 nope:(
1 5 0
```
(exact answers will vary depending on how exactly the pid namespace is created, of course.)

I strongly recommend not doing a straight revert. Do revert the wineserver change, but reduce the ntdll change to a warning (wine's ERR macro, for example). Don't change wineserver; instead, change ntdll's getpid() to `(server_pid > 0 ? getpid() : -1)`.

That should make everything needing a PID return an error. Still broken, of course, but better than returning nonsense, like the caller using zero bytes of RAM.

(Full disclosure: I wrote that MR)

**Comment by KOPRajs** [comment]
It would be best to find a way how to run the second application in the already running container instance. It is technically possible using the `flatpak enter`, we just need to find a way how to do it nicely.

**Comment by KOPRajs** [comment]
> Why not treat Wine as a runtime, or a base application upon which flatpak apps can be built? This way, each app would have its own wine prefix and wineserver instance.

You would still run into this issue when you run multiple instances of the app.

I use the Wine in Flatpak to run Windows apps just like I would use Wine without Flatpak. The motivation is mostly not the sandboxing, but to avoid having to install 32-bit deps on the host system.

**Comment by Erick555** [comment]
> I agree, but it would greatly limit the usability of the Wine in Flatpak. We only have single prefix, which means to only run one app at a time (even only one instance of the same app).

What do you mean? You can add `--env=WINEPREFIX=foo` in cmdline or desktop shortcut and run as many apps as you want. It's not different than outside flatpak.

**Comment by foresto** [comment]
> It would be best to find a way how to run the second application in the already running container instance.

I don't think that can reasonably be considered "best", since it breaks the flatpak model of isolating applications from each other and breaks their separation from runtimes. (Wine fundamentally is a runtime environment, after all.)

I can accept that it might be desirable for certain use cases, though.

~~Maybe flatpak extension points could be helpful in getting the behavior you want? (I haven't packaged with them, so I can't offer guidance here.)~~ Edit: I suspect extension points would be too application-specific.

**Comment by Alcaro** [comment]
If you're running four different programs in four different wineprefixes, then yes, they're supposed to be isolated from each other.

If you're running four instances of the same program in the same prefix, then you probably want them able to see each other, and live in the same container.

To me, both of those sound like legitimate usecases. (Though I don't know enough about Flatpak to propose any sane way to support both simultaneously.)

**Comment by KOPRajs** [comment]
> To me, both of those sound like legitimate usecases. (Though I don't know enough about Flatpak to propose any sane way to support both simultaneously.)

Using `flatpak enter` when you want to use the Wine in flatpak as a standalone application, while building an app specific flatpaks for applications, that are supposed to run isolated in their own container and using their own prefix, seems reasonable. Or providing a custom prefix for a specific app in launch options.

**Comment by user1-github** [comment]
To me this seems like something worth reporting to [flatpak/issues](https://github.com/flatpak/flatpak/issues) especially now when its development has sped up lately.
